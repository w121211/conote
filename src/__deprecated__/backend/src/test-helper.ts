import dotenv from 'dotenv'
import omitDeep from 'omit-deep-lodash'
import { removeUndefinedFields } from '../../lib/editor/src/test-helper'

export function _clean(obj: any): any {
  return removeUndefinedFields(omitDeep(obj, ['createdAt', 'updatedAt']))
}

const config = dotenv.config()
if (config.error) throw config.error
if (!config.parsed?.BOT_EMAIL || !config.parsed?.BOT_PASSWORD) {
  throw new Error('BOT_EMAIL or BOT_PASSWORD not found in .env')
}

export const BOT = { id: 'bot', email: config.parsed.BOT_EMAIL, password: config.parsed.BOT_PASSWORD }

export const TESTUSERS = [
  { id: 'test-user-1', email: 'aaa@aaa.com', password: 'aaa' },
  { id: 'test-user-2', email: 'bbb@bbb.com', password: 'bbb' },
  { id: 'test-user-3', email: 'ccc@ccc.com', password: 'ccc' },
  { id: 'test-user-4', email: 'ddd@ddd.com', password: 'ddd' },
  { id: 'test-user-5', email: 'eee@eee.com', password: 'eee' },
]

export const TESTOAUTHORS = [{ name: 'test-oauthor-1' }]
