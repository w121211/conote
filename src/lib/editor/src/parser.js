"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.splitByUrl = exports.findUrl = exports.tokenizeSection = exports.tokenizeSymbol = void 0;
const prismjs_1 = __importDefault(require("prismjs"));
const helper_1 = require("./helper");
const GRAMMAR = {
    'multiline-marker': {
        // pattern: /^\[[^\s\]]*\]$(?:\n^(?!\[).+)+/m,
        pattern: /^\[[^\s[\]]*\]$(?:\n^(?!\[[^\s[\]]*\]).+)+/m,
        inside: {
            'line-mark': {
                pattern: /^\[[^\s[\]]*\]$/m,
            },
            'line-value': {
                pattern: /^.+$/m,
                inside: {
                    ticker: { pattern: /\$[A-Z-]+/ },
                    topic: { pattern: /\[\[[^\]]+\]\]/u },
                    stamp: { pattern: /\s%[a-zA-Z0-9]{3}$/ },
                },
            },
            // 'list-string': {
            //     pattern: /^[-]+\s.+$/m,
            // },
        },
    },
    'inline-marker': {
        pattern: /^\[[^\s[\]]+\].*$/m,
        inside: {
            'inline-mark': {
                pattern: /^\[[^\s[\]]+\]/,
            },
            'inline-value': {
                pattern: /^.+$/,
                inside: {
                    ticker: { pattern: /\$[A-Z-]+/ },
                    topic: { pattern: /\[\[[^\]]+\]\]/u },
                    // TODO: 沒辦法將space與stamp分開（會成為一個string，需要trim）
                    stamp: { pattern: /\s%[a-zA-Z0-9]{3}$/ },
                },
            },
        },
    },
    ticker: { pattern: /\$[A-Z-]+/ },
    topic: { pattern: /\[\[[^\]]+\]\]/u },
    // 'radio': {
    //   pattern: /[\s\t]+\[\w?\]\p{L}+/u,
    //   // greedy: true
    // },
};
const SYMBOL_GRAMMAR = {
    ticker: GRAMMAR.ticker,
    topic: GRAMMAR.topic,
};
const SECTION_GRAMMAR = {
    'sect-ticker': {
        alias: 'sect-ticker',
        pattern: /^\n\$[A-Z-]+(@\w+)?$/m,
        inside: {
            'sect-symbol': { pattern: /^\n\$[A-Z-]+/ },
            'sect-user': { pattern: /@\w+/ },
        },
    },
    'sect-ticker-begin-line': {
        alias: 'sect-ticker',
        pattern: /^\$[A-Z-]+(@\w+)?\n/,
        inside: {
            'sect-symbol': { pattern: /^\$[A-Z-]+/ },
            'sect-user': { pattern: /@\w+/ },
        },
    },
    'sect-topic': {
        alias: 'sect-topic',
        pattern: /^\n\[\[[^\]]+\]\](@\w+)?$/m,
        inside: {
            'sect-symbol': { pattern: /^\n\[\[[^\]]+\]\]/u },
            'sect-user': { pattern: /@\w+/ },
        },
    },
    'sect-topic-begin-line': {
        alias: 'sect-topic',
        pattern: /^\[\[[^\]]+\]\](@\w+)?\n/,
        inside: {
            'sect-symbol': { pattern: /^\[\[[^\]]+\]\]/u },
            'sect-user': { pattern: /@\w+/ },
        },
    },
    'sect-breaker': {
        alias: 'sect-breaker',
        pattern: /^\n\/{3,}.+$/m,
    },
    'sect-url': {
        alias: 'sect-url',
        pattern: /^https?:\/\/(www\.)?[-a-zA-Z0-9@:%._+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b([-a-zA-Z0-9()@:%_+.~#?&//=]*)$/m,
    },
};
function tokenizeSymbol(text) {
    /** 將text中的symbol($AA, [[Topic]])轉成token */
    return prismjs_1.default.tokenize(text, SYMBOL_GRAMMAR);
}
exports.tokenizeSymbol = tokenizeSymbol;
function validate(items, allowedMarkers) {
    for (const e of items) {
        const format = allowedMarkers.find(f => f.mark === e.mark);
        if (format === undefined) {
            e.error = '不在允許的markers中';
            continue;
        }
        if (e.value === undefined) {
            e.error = '沒有value';
            continue;
        }
        if (format.validater && !format.validater(e.value)) {
            e.error = 'value不符合格式';
            continue;
        }
        if (format.inline) {
            const filtered = items.filter(e => e.mark === format.mark);
            for (let i = 0; i < filtered.length - 1; i++) {
                filtered[i].error = '只能define一次，最後一個會被保留';
            }
        }
        // if (e.mark === MARKER_FORMAT.card.mark && getCardId) {
        //     getCardId(e.value)
        //         .then(result => {
        //             if (result === null) {
        //                 e.error = '找不到對應的card'
        //             } else {
        //                 e.cardId = result
        //             }
        //         })
        // }
        // if (e.children) {
        //     if (format.nested) {
        //         // TODO: 當前只有ticker可以nested，應該要靈活一些
        //         e.children = validate(e.children, TICKER_FORMATTER)
        //     } else {
        //         e.error = '不允許nested'
        //     }
        // }
    }
    return items;
}
function tokenizeSection(text, 
// symbolCardDict: Record<string, CardIdentifier>,
// nestedCards: CardIdentifier[] = [],
rootFormat, oauthorName, allowedSects = ['ticker', 'topic']) {
    // 分出text中的每個section
    // let _sect: {bodyTokens: []} = { bodyTokens: [] }
    var _a, _b;
    function _parseSectToken(token) {
        let symbol;
        let user;
        if (Array.isArray(token.content)) {
            for (const e of token.content) {
                if (typeof e === 'string') {
                    // do nothing
                }
                else if (e.type === 'sect-symbol') {
                    symbol = helper_1.streamToStr(e.content).trim();
                }
                else if (e.type === 'sect-user') {
                    user = helper_1.streamToStr(e.content);
                }
            }
            if (symbol === undefined) {
                console.error(token);
                throw new Error();
            }
            return { symbol, user };
        }
        console.error(token);
        throw new Error();
    }
    const sects = [];
    let _sect = {
        bodyTokens: [],
    };
    // let _sectToken: Prism.Token
    // let _bodyTokens: (Prism.Token | string)[] = []
    for (const e of prismjs_1.default.tokenize(text, SECTION_GRAMMAR)) {
        if (typeof e === 'string') {
            _sect.bodyTokens.push(e);
        }
        else if (e.alias === 'sect-ticker') {
            // 先儲存之前的section
            sects.push(Object.assign({}, _sect));
            // TODO: @ME
            const parsed = _parseSectToken(e);
            _sect = {
                ticker: true,
                nestedCard: { symbol: parsed.symbol, oauthor: (_a = parsed.user) !== null && _a !== void 0 ? _a : oauthorName },
                sectToken: e,
                bodyTokens: [],
            };
        }
        else if (e.alias === 'sect-topic') {
            sects.push(Object.assign({}, _sect));
            const parsed = _parseSectToken(e);
            _sect = {
                topic: true,
                nestedCard: { symbol: parsed.symbol, oauthor: (_b = parsed.user) !== null && _b !== void 0 ? _b : oauthorName },
                sectToken: e,
                bodyTokens: [],
            };
        }
        else if (e.alias === 'sect-breaker') {
            sects.push(Object.assign({}, _sect));
            // 把breaker視為獨立的section
            sects.push({ breaker: true, sectToken: e, bodyTokens: [] });
            // breaker之後的新section
            _sect = { bodyTokens: [] };
        }
        else {
            console.error(e);
            throw new Error('應該要處理但未處理的token-type');
        }
    }
    sects.push(Object.assign({}, _sect));
    for (const e of sects) {
        const body = helper_1.streamToStr(e.bodyTokens);
        if (e.ticker || e.topic) {
            // 對section做tokenize section-body
            e.bodyTokens = prismjs_1.default.tokenize(body, GRAMMAR);
        }
        else if (!e.breaker && body.length > 0) {
            // 所有非ticker、topic的card & 有body-string的section皆視為root
            e.root = true;
            e.bodyTokens = prismjs_1.default.tokenize(body, GRAMMAR);
        }
    }
    // 對每個token紀錄linenumber、marker
    let linenumber = 0;
    let mark = null;
    function _recursiveExtend(stream) {
        if (typeof stream === 'string') {
            linenumber += stream.split('\n').length - 1;
            return stream;
        }
        else if (Array.isArray(stream)) {
            return stream.map(e => _recursiveExtend(e));
        }
        else if (stream.type === 'inline-mark' || stream.type === 'line-mark') {
            mark = helper_1.streamToStr(stream.content);
            if (mark === '') {
                throw new Error();
            }
            return Object.assign(Object.assign({}, stream), { linenumber, marker: { mark }, content: _recursiveExtend(stream.content) });
        }
        else if (stream.type === 'inline-value' || stream.type === 'line-value') {
            const value = helper_1.streamToStr(stream.content, 'stamp').trim();
            if (mark === null) {
                throw new Error();
            }
            return Object.assign(Object.assign({}, stream), { linenumber, marker: { mark, value }, content: _recursiveExtend(stream.content) });
        }
        else {
            return Object.assign(Object.assign({}, stream), { linenumber, content: _recursiveExtend(stream.content) });
        }
    }
    for (const e of sects) {
        // section的全部stream（含section-token, body-tokens)
        let stream = [];
        if (e.sectToken)
            stream.push(e.sectToken);
        stream = stream.concat(e.bodyTokens);
        // 轉成ext-token
        mark = null;
        e.stream = _recursiveExtend(stream);
    }
    return sects.map(e => {
        return {
            root: e.root,
            breaker: e.breaker,
            ticker: e.ticker,
            topic: e.topic,
            nestedCard: e.nestedCard,
            stream: e.stream,
        };
    });
}
exports.tokenizeSection = tokenizeSection;
function findUrl(text) {
    /** 從首先出現的單行URL開始，擷取從URL以下的文章 */
    const tokens = prismjs_1.default.tokenize(text, SECTION_GRAMMAR);
    let url;
    const _tokensAfterUrl = [];
    for (const e of tokens) {
        if (url === undefined) {
            if (typeof e !== 'string' && e.type === 'sect-url')
                url = e.content;
            else
                continue;
        }
        _tokensAfterUrl.push(e);
    }
    // return { url, textAfterUrl: tokensToText(_tokensAfterUrl) }
    return { url, textAfterUrl: 'tokensToText(_tokensAfterUrl)' };
}
exports.findUrl = findUrl;
function splitByUrl(text) {
    /** 以text中的單行URL做split，返回:[url, part-text][] */
    const tokens = prismjs_1.default.tokenize(text, SECTION_GRAMMAR);
    let buffer = [];
    let url;
    const splits = [];
    for (const e of tokens) {
        if (typeof e !== 'string' && e.type === 'sect-url') {
            // 儲存前一個buffer後清空
            splits.push([url, helper_1.streamToStr(buffer)]);
            buffer = [];
            // 當前url
            url = e.content;
        }
        else {
            buffer.push(e);
        }
    }
    // 儲存最後一個
    splits.push([url, helper_1.streamToStr(buffer)]);
    return splits;
}
exports.splitByUrl = splitByUrl;
