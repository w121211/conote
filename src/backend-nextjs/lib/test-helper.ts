import dotenv from 'dotenv'
import { hash, hashSync } from 'bcryptjs'
import { PrismaClient, SymbolCat } from '@prisma/client'
// import omitDeep from 'omit-deep-lodash'
import { omitUndefined } from '../../lib/editor/src/test-helper'
<<<<<<< HEAD
=======

>>>>>>> backend-dev
/**
 * See: https://github.com/lodash/lodash/issues/723
 * Recursively remove keys from an object
 * @param {object} input
 * @param {Array<number | string>>} excludes
 * @return {object}
 */
function omitDeep(input: Record<string, unknown>, excludes: Array<number | string>): Record<string, unknown> {
  return Object.entries(input).reduce<Record<string, unknown>>((acc, [key, value]) => {
    const shouldExclude = excludes.includes(key)
    if (shouldExclude) return acc

    if (Array.isArray(value)) {
      const arrValue = value
      const nextValue = arrValue.map(arrItem => {
        if (typeof arrItem === 'object') {
          return omitDeep(arrItem, excludes)
        }
        return arrItem
      })
      acc[key] = nextValue
      return acc
    } else if (typeof value === 'object') {
      acc[key] = omitDeep(value as Record<string, unknown>, excludes)
      return acc
    }

    acc[key] = value

    return acc
  }, {})
}

<<<<<<< HEAD
export function clean(obj: Record<string, unknown>): Record<string, unknown> {
  return omitUndefined(omitDeep(obj, ['createdAt', 'updatedAt']))
=======
export function clean(obj: Record<string, unknown> | null): Record<string, unknown> | null {
  return obj === null ? obj : omitUndefined(omitDeep(obj, ['createdAt', 'updatedAt']))
>>>>>>> backend-dev
}

const config = dotenv.config()

if (config.error) throw config.error

if (!config.parsed?.BOT_EMAIL || !config.parsed?.BOT_PASSWORD) {
  throw new Error('BOT_EMAIL or BOT_PASSWORD not found in .env')
}

export const BOT = { id: 'bot', email: config.parsed.BOT_EMAIL, password: config.parsed.BOT_PASSWORD }

export const TESTUSERS = [
  { id: 'test-user-1', email: 'aaa@aaa.com', password: 'aaa' },
  { id: 'test-user-2', email: 'bbb@bbb.com', password: 'bbb' },
  { id: 'test-user-3', email: 'ccc@ccc.com', password: 'ccc' },
  { id: 'test-user-4', email: 'ddd@ddd.com', password: 'ddd' },
  { id: 'test-user-5', email: 'eee@eee.com', password: 'eee' },
]

export const TESTOAUTHORS = [{ name: 'test-oauthor-1' }]

export const TEST_SYMBOLS = [
  { name: '$AAA', cat: SymbolCat.TICKER },
  { name: '$ABB', cat: SymbolCat.TICKER },
  { name: '$ACC', cat: SymbolCat.TICKER },
  { name: '$BBB', cat: SymbolCat.TICKER },
  { name: '$CCC', cat: SymbolCat.TICKER },
  { name: '$DDD', cat: SymbolCat.TICKER },
  { name: '[[Apple]]', cat: SymbolCat.TOPIC },
  { name: '[[Google]]', cat: SymbolCat.TOPIC },
  { name: '[[蘋果]]', cat: SymbolCat.TOPIC },
  { name: '[[估狗]]', cat: SymbolCat.TOPIC },
  { name: '[[Apple love Google]]', cat: SymbolCat.TOPIC },
  { name: '[[Google hate Apple]]', cat: SymbolCat.TOPIC },
]

export async function createTestUsers(prisma: PrismaClient): Promise<void> {
  await prisma.user.create({
    data: { id: BOT.id, email: BOT.email, password: await hash(BOT.password, 10) },
  })
  await prisma.$transaction(
    TESTUSERS.map(e =>
      prisma.user.create({
        data: { id: e.id, email: e.email, password: hashSync(e.password, 10) },
      }),
    ),
  )
  await prisma.$transaction(
    TESTOAUTHORS.map(e =>
      prisma.oauthor.create({
        data: { name: e.name },
      }),
    ),
  )
}

export async function createTestSymbols(prisma: PrismaClient): Promise<void> {
  await prisma.$transaction(
    TEST_SYMBOLS.map(e =>
      prisma.symbol.create({
        data: { name: e.name, cat: e.cat },
      }),
    ),
  )
}
