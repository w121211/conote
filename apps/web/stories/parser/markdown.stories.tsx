import React from 'react'
import { ComponentMeta, ComponentStory } from '@storybook/react'
import ReactMarkdown from 'react-markdown'
import { ReactMarkdownOptions } from 'react-markdown/lib/react-markdown'
import ParseRenderEl from '../../frontend/components/editor-textarea/src/components/inline/parse-render-el'

export default {
  title: 'parsers/ReactMarkdown',
  component: ReactMarkdown,
} as ComponentMeta<typeof ReactMarkdown>

// function rehypeWrapText(): Plugin {
//   return (tree: Root) => {
//     // visitParents(tree, 'text', (node, ancestors) => {
//     //   if (ancestors.at(-1).tagName !== 'custom-typography') {
//     //     node.type = 'element'
//     //     node.tagName = 'custom-typography'
//     //     node.children = [{ type: 'text', value: node.value }]
//     //   }
//     // })
//     visit(tree, 'text', (node, i, parent) => {
//       console.log(node, parent)
//     })
//   }
// }

export const Base = () => {
  const markdown = `A paragraph with *emphasis* and **strong importance**. [[hello world]]

#this is something cool-123456789#

#Which browser is best at keeping things confidential?-mock_discuss_0_active#

> A block quote with ~strikethrough~ and a URL: https://reactjs.org. [[hello world]]

* Lists
* [ ] todo
* [x] done

A table:

| a | b |
| - | - |
`

  const components: ReactMarkdownOptions['components'] = {
    p: ({ node, children, ...props }) => {
      return (
        <p {...props}>
          {children.map((e, i) => {
            if (typeof e === 'string') {
              return (
                <span key={i}>
                  <ParseRenderEl key={i} blockUid="blockUid" str={e} isViewer />
                </span>
              )
            } else return e
          })}
        </p>
      )
    },
  }

  return <ReactMarkdown components={components}>{markdown}</ReactMarkdown>
}
