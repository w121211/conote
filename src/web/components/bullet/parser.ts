import { Grammar, Token, tokenize as prismTokenize } from 'prismjs'
import { ShotChoice } from 'graphql-let/__generated__/__types__'
// import { parseInlineShotParams } from '../models/shot'
import { tokenToString } from '../../lib/token'
import { InlineItem } from './types'

/**
 * Regex:
 * - url @see https://stackoverflow.com/questions/3809401/what-is-a-good-regular-expression-to-match-a-url
 * - hashtag @see https://stackoverflow.com/questions/38506598/regular-expression-to-match-hashtag-but-not-hashtag-with-semicolon
 */
const reTicker = /\$[A-Z-=]+/
const reTopic = /\[\[[^\]\n]+\]\]/u
export const reAuthor = /\B@[\p{L}\d_]+\b/u

const decorateReMirrorTicker = /^(::\$[A-Z-=]+)\b/u
const decorateReMirrorTopic = /^(::\[\[[\p{Letter}\d\s(),-]+\]\])\B/u
const reMirrorTicker = /^(::\$[A-Z-=]+)\b(?:\s@([\p{Letter}\d_]+))?/u
const reMirrorTopic = /^(::\[\[[\p{Letter}\d\s(),-]+\]\])\B(?:\s@([\p{Letter}\d_]+))?/u
// const rePoll = /\B!\(\(poll:(\d+)\)\)\(((?:#[a-zA-Z0-9]+\s)+#[a-zA-Z0-9]+)\)\B/
const rePoll = /\B!\(\(poll:(c[a-z0-9]{24,29})\)\)\(((?:#[a-zA-Z0-9]+\s)+#[a-zA-Z0-9]+)\)\B/
const reNewPoll = /\B!\(\(poll\)\)\(((?:#[a-zA-Z0-9]+\s)+#[a-zA-Z0-9]+)\)\B/
const reShot = /\B!\(\(shot:(c[a-z0-9]{24,29})\)\)\(([^)]*)\)\B/
const reNewShot = /\B!\(\(shot\)\)\(([^)]*)\)\B/

const grammar: Grammar = {
  // 順序重要，先 mirror 後 ticker
  'mirror-ticker': { pattern: reMirrorTicker },
  'mirror-topic': { pattern: reMirrorTopic },

  poll: { pattern: rePoll },
  'new-poll': { pattern: reNewPoll },

  shot: { pattern: reShot },
  'new-shot': { pattern: reNewShot },

  ticker: { pattern: reTicker },
  topic: { pattern: reTopic },

  filtertag: { pattern: /(?<=\s|^)#[a-zA-Z0-9()]+(?=\s|$)/ },

  url: {
    pattern: /(?<=\s|^)@(https?:\/\/(?:www\.|(?!www))[a-zA-Z0-9][a-zA-Z0-9-]+[a-zA-Z0-9]\.[^\s]{2,})(?=\s|$)/,
  },

  author: { pattern: reAuthor },
}

const decorationGrammar: Grammar = {
  // 'mirror-ticker': { pattern: /^::\$[A-Z-=]+\b/ },
  // 'mirror-topic': { pattern: /^::\[\[[^\]\n]+\]\]\B/u },
  'mirror-ticker': { pattern: decorateReMirrorTicker, inside: { punctuation: /^::|\[\[|\]\]/g } },
  'mirror-topic': { pattern: decorateReMirrorTopic, inside: { punctuation: /^::|\[\[|\]\]/g } },
  poll: { pattern: rePoll },
  'new-poll': { pattern: reNewPoll },
  shot: { pattern: reShot },
  'new-shot': { pattern: reNewShot },
  ticker: { pattern: reTicker },
  topic: { pattern: reTopic, inside: { punctuation: /^::|\[\[|\]\]/g } },
  filtertag: { pattern: /(?<=\s|^)#[a-zA-Z0-9()]+(?=\s|$)/ },
  url: {
    pattern: /(?<=\s|^)@(https?:\/\/(?:www\.|(?!www))[a-zA-Z0-9][a-zA-Z0-9-]+[a-zA-Z0-9]\.[^\s]{2,})(?=\s|$)/,
  },
  author: { pattern: reAuthor },
}

export function decorationTokenizer(text: string): (string | Token)[] {
  return prismTokenize(text, decorationGrammar)
}
export function tokenizeBulletString(text: string): (string | Token)[] {
  return prismTokenize(text, grammar)
}

export function inlinesToString(inlines: InlineItem[]): string {
  return inlines.reduce((acc, cur) => `${acc}${cur.str}`, '')
}

const isShotChoice = (s: string): s is ShotChoice => {
  return ['#LONG', '#SHORT', '#HOLD'].includes(s)
}

const parseInlineShotParams = (
  params: string[],
): {
  authorName?: string
  targetSymbol?: string
  choice?: ShotChoice
} => {
  const [_authorName, _targetSymbol, _choice] = params
  // const matchAuthor = reAuthor.exec(_authorName)
  // const matchSymbol = parseSymbol(_targetSymbol)
  return {
    authorName: _authorName,
    targetSymbol: _targetSymbol,
    choice: isShotChoice(_choice) ? _choice : undefined,
  }
}

export const BulletParser = {
  /**
   * Parse a bullet head/body string to slate inline elements
   *
   * @throws Parse error
   */
  parseBulletHead({ str }: { str: string }): { inlines: InlineItem[] } {
    // TODO: validate
    function _tokenToInline(token: string | Token): InlineItem {
      if (typeof token === 'string') {
        return { type: 'text', str: token }
      }

      const str = tokenToString(token.content)
      switch (token.type) {
        case 'mirror-ticker':
        case 'mirror-topic': {
          const match = token.type === 'mirror-ticker' ? reMirrorTicker.exec(str) : reMirrorTopic.exec(str)
          if (match) {
            return {
              type: 'mirror',
              str,
              mirrorSymbol: match[1],
              author: match[2], // 沒有 match 到時會返回 undefined
            }
          } else {
            console.error(str)
            throw 'Parse error'
          }
        }
        case 'filtertag': {
          return { type: 'filtertag', str }
        }
        case 'url':
        case 'ticker':
        case 'topic':
        case 'author': {
          return { type: 'symbol', str, symbol: str }
        }
        case 'poll': {
          const match = rePoll.exec(str)
          if (match) {
            return {
              type: 'poll',
              str,
              id: match[1],
              choices: match[2].split(' '),
            }
          } else {
            console.error(str)
            throw 'Parse error'
          }
        }
        case 'new-poll': {
          const match = reNewPoll.exec(str)
          if (match) {
            return {
              type: 'poll',
              str,
              choices: match[1].split(' '),
            }
          } else {
            console.error(str)
            throw 'Parse error'
          }
        }
        case 'shot': {
          const match = reShot.exec(str)
          if (match) {
            console.log(match[2])
            const params = match[2].split(' ')
            const { authorName, targetSymbol, choice } = parseInlineShotParams(params)
            return {
              type: 'shot',
              id: match[1],
              str,
              params,
              authorName,
              targetSymbol,
              choice,
            }
          } else {
            console.error(str)
            throw 'Parse error'
          }
        }
        case 'new-shot': {
          const match = reNewShot.exec(str)
          if (match) {
            const params = match[1].split(' ')
            const { authorName, targetSymbol, choice } = parseInlineShotParams(params)
            return {
              type: 'shot',
              str,
              params,
              authorName,
              targetSymbol,
              choice,
            }
          } else {
            console.error(str)
            throw 'Parse error'
          }
        }
      }
      // 沒被處理到的 token
      console.error(token)
      throw 'Parse error'
    }

    const tokens = tokenizeBulletString(str)
    const inlines = tokens.map(e => _tokenToInline(e))
    return { inlines }
  },
}
