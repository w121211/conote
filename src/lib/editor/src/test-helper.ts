import { readFileSync } from 'fs'
import { splitByUrl } from './parser'

export function load(filepath: string): [string, string][] {
  return (
    splitByUrl(readFileSync(filepath, { encoding: 'utf8' }))
      .filter((e): e is [string, string] => e[0] !== undefined)
      // .map(e => [e[0], `\n${e[1].trim()}`])
      .map(e => [e[0], e[1].trim()])
  )
}

export class FakeChance {
  i = 0
  string(): string {
    const str = this.i.toString(36).padStart(3, '0')
    this.i += 1
    if (this.i >= 36 * 36 * 36) {
      this.i = 0
    }
    return str
  }
}

export function removeUndefinedFields<T>(obj: T): T {
  return JSON.parse(JSON.stringify(obj))
}
