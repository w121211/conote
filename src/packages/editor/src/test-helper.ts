import { readFileSync } from 'fs'
import { splitByUrl } from './parser'

export function load(filepath: string): [string, string][] {
  return (
    splitByUrl(readFileSync(filepath, { encoding: 'utf8' }))
      .filter((e): e is [string, string] => e[0] !== undefined)
      // .map(e => [e[0], `\n${e[1].trim()}`])
      .map(e => [e[0], e[1].trim()])
  )
}

export class FakeChance {
  private i = 0

  public string(): string {
    const str = this.i.toString(36).padStart(3, '0')
    this.i += 1
    if (this.i >= 36 * 36 * 36) {
      this.i = 0
    }
    return str
  }

  public reset(): void {
    this.i = 0
  }
}

export function omitUndefined<T>(obj: T): T {
  return JSON.parse(JSON.stringify(obj))
}
