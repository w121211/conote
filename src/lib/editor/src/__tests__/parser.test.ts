import { resolve } from 'path'
import { splitByUrl, tokenizeSection } from '../parser'
import { load } from '../test-helper'

describe('Parser', () => {
  it('use url to split text to sections', () => {
    const text = [
      'https://www.youtube.com/watch?v=qSYGlOZNUCw\n',
      '$AAAA\n\n',
      'https://www.youtube.com/watch?v=kprnW5VadnY&list=WL&index=53\n',
      '$BBBB\n',
    ].join('')
    expect(splitByUrl(text)).toEqual([
      [undefined, ''],
      ['https://www.youtube.com/watch?v=qSYGlOZNUCw', '\n$AAAA\n\n'],
      ['https://www.youtube.com/watch?v=kprnW5VadnY&list=WL&index=53', '\n$BBBB\n'],
    ])
  })

  it('tokenize inline-marker', () => {
    const a = '[=] [[大數據]] [[SaaS]]'
    expect(tokenizeSection(a)[0].stream).toEqual([
      {
        type: 'inline-marker',
        alias: undefined,
        length: 20,
        linenumber: 0,
        content: [
          {
            type: 'inline-mark',
            alias: undefined,
            content: '[=]',
            length: 3,
            linenumber: 0,
            marker: { key: '[=]' },
          },
          {
            type: 'inline-value',
            alias: undefined,
            length: 17,
            linenumber: 0,
            marker: {
              key: '[=]',
              value: '[[大數據]] [[SaaS]]',
            },
            content: [
              ' ',
              {
                type: 'topic',
                alias: undefined,
                content: '[[大數據]]',
                length: 7,
                linenumber: 0,
              },
              ' ',
              {
                type: 'topic',
                alias: undefined,
                content: '[[SaaS]]',
                length: 8,
                linenumber: 0,
              },
            ],
          },
        ],
      },
    ])
  })

  it.each<[string, string]>(load(resolve(__dirname, '__samples', 'common.txt')))(
    'tokenize to sections',
    (url: string, body: string) => {
      const sects = tokenizeSection(body).map(e => ({
        root: e.root,
        breaker: e.breaker,
        ticker: e.ticker,
        topic: e.topic,
        nestedCard: e.nestedCard,
      }))
      expect(sects).toMatchSnapshot()
    },
  )
})
