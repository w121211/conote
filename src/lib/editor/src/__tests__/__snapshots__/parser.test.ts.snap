// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser tokenize to sections 1`] = `
Array [
  Object {
    "breaker": undefined,
    "nestedCard": undefined,
    "root": true,
    "ticker": undefined,
    "topic": undefined,
  },
]
`;

exports[`Parser tokenize to sections 2`] = `
Array [
  Object {
    "breaker": undefined,
    "nestedCard": undefined,
    "root": undefined,
    "ticker": undefined,
    "topic": undefined,
  },
  Object {
    "breaker": undefined,
    "nestedCard": Object {
      "oauthor": undefined,
      "symbol": "$AAA",
    },
    "root": undefined,
    "ticker": true,
    "topic": undefined,
  },
  Object {
    "breaker": undefined,
    "nestedCard": Object {
      "oauthor": undefined,
      "symbol": "[[TOPIC]]",
    },
    "root": undefined,
    "ticker": undefined,
    "topic": true,
  },
  Object {
    "breaker": undefined,
    "nestedCard": Object {
      "oauthor": undefined,
      "symbol": "$BBB",
    },
    "root": undefined,
    "ticker": true,
    "topic": undefined,
  },
]
`;
